kernels/lgamma/lgamma.mlu

```diff
@@ -0,0 +1,273 @@
+  __bang_bnot((char *)sel, (char *)sel, sz * sizeof(float));
+  __bang_band((char *)sel, (char *)b, (char *)sel, sz * sizeof(float));
+  __bang_add(dst, dst, sel, sz);
+}
+
+__mlu_func__ void bang_isfinite(float *dst, float *src, uint32_t sz) {
+  __bang_band_scalar(dst, src, (float)inf_float_value, sz);
+  __bang_ne_scalar(dst, dst, (float)inf_float_value, sz);
+}
+
+__mlu_func__ void bang_isinf(float *dst, float *src, uint32_t sz) {
+  __bang_band_scalar(dst, src, (float)nan_float_mask, sz);
+  __bang_eq_scalar(dst, dst, (float)inf_float_value, sz);
+}
+
+__mlu_func__ void __bang_loghp(float *dst, float *src, uint32_t sz) {
+  __bang_log(dst, src, sz);
+  __bang_mul_scalar(dst, dst, log_e_2, sz);
+}
+
+// template <typename T>
+__mlu_func__ void calc_lgamma(float *buf0, float *buf1, float *buf2,
+                              float *buf3, float *buf4, int num_deal) {
+  static const int numCoeff = 4;
+  static const float log_pi = 1.144730;
+  static const float coeff[numCoeff] = {
+      2.50663,
+      34.7892,
+      -20.8994,
+      1.35914,
+  };
+
+  /**
+   * bool need_to_reflect = (input < 0);
+   */
+  __bang_lt_scalar(buf4, buf0, 0, num_deal);
+
+  /**
+   * float reflect_x = need_reflect ? -input : input - 1;
+   */
+  // using buf1 -> reflect_x
+  __bang_mul_scalar(buf1, buf0, -1, num_deal);
+  __bang_add_scalar(buf2, buf0, -1, num_deal);
+  bang_mux2(buf1, buf1, buf2, buf4, num_deal);
+
+  /**
+    float accm = coeffs[0];
+    int numCoeff = coeffs.size();   // aka a
+    for (size_t k = 1; k < numCoeff; k++) {
+      accm += coeffs[k] / (reflect_x + k);
+    }
+    */
+  // do loop flatten, using buf2 -> sum
+  __bang_add_scalar(buf3, buf1, 1, num_deal);
+  __bang_recip(buf3, buf3, num_deal);
+  __bang_fusion(FUSION_FMA, buf2, buf3, coeff[1], coeff[0], num_deal);
+
+  __bang_add_scalar(buf3, buf1, 2, num_deal);
+  __bang_recip(buf3, buf3, num_deal);
+  __bang_fusion(FUSION_FMA, buf2, buf3, coeff[2], buf2, num_deal, num_deal);
+
+  __bang_add_scalar(buf3, buf1, 3, num_deal);
+  __bang_recip(buf3, buf3, num_deal);
+  __bang_fusion(FUSION_FMA, buf2, buf3, coeff[3], buf2, num_deal, num_deal);
+
+  /**
+   * float lgamma_x = (reflect_x+0.5)*log(reflect_x+numCoeff) -
+   * (reflect_x+numCoeff) + log(accm);
+   */
+  // using buf1 -> lgamma_x
+  __bang_add_scalar(buf3, buf1, numCoeff, num_deal);
+  __bang_add_scalar(buf1, buf1, 0.5, num_deal);
+  __bang_loghp(buf2, buf2, num_deal);
+  __bang_sub(buf2, buf2, buf3, num_deal);
+  __bang_loghp(buf3, buf3, num_deal);
+  __bang_fusion(FUSION_FMA, buf1, buf3, buf1, buf2, num_deal, num_deal);
+
+  /**
+   * float abs_input = fabs(input);
+   * float abs_frac_input = abs_input - floorf(abs_input);
+   */
+  // using buf4 -> abs_frac_input
+  __bang_abs(buf2, buf0, num_deal);
+  __bang_floor(buf3, buf2, num_deal);
+  __bang_sub(buf4, buf2, buf3, num_deal);
+
+  /**
+   * float reduced_frac_input = (abs_frac_input > 0.5) ? 1 - abs_frac_input :
```

Comment:
这里是为了精度，可以看下原本实现中代码注释部分：
   * We don't have a frac(x) primitive in XLA and computing it is tricky, but
   * because abs(sin(pi * x)) = abs(sin(pi * abs(x))), it's good enough for
   * our purposes to use abs(frac(x)) = abs(x) - floor(abs(x)).
   *
   * Furthermore, pi * abs(frac(x)) loses precision when abs(frac(x)) is close
   * to 1.  To remedy this, we can use the fact that sin(pi * x) in the domain
   * [0, 1] is symmetric across the line Y=0.5.

Reply:
ok